### ISSUE TITLE
How to send HTTPS request to the web site that uses SSL Certificate (using a password to load)?

### ISSUE THREAD

**Issue by walterdev** [openresty/openresty#222]
Using openresty to do the forward proxy server.
When sending HTTP request and no certificate HTTPS request, use the resty.http , but can not send request with SSL certificate.
lua_ssl_trusted_certificate interface should load the SSL certificate, but seemingly can not be loaded with a SSL certificate with password.
What can I do?

**Comment by detailyang** [comment]
Hello.

@walterdev there is a workaround to solve it that just run `openssl rsa -in password.pem -out nopassword.pem`. wish it can help you:)

**Comment by walterdev** [comment]
Hello @detailyang
I have four files of certificate.
apiclient_cert.p12
apiclient_cert.pem
apiclient_key.pem
rootca.pem

I want to use the certificates to connect the third website api , like "https://api.mch.weixin.qq.com/secapi/pay/refund".

Please tell me what can I do step by step. thanks a lot.

**Comment by detailyang** [comment]
Hello

@walterdev , i guess you are debuging api for wechat:). I thinks the apiclient_key.pem is the private key issued by wechat, so you can try this to generate unsecure key by `openssl rsa -in apiclient_key.pem -out apiclient_key.pem.unsecure`

**Comment by walterdev** [comment]
Hello

@detailyang , thanks for your help.
I can use the command to make a apiclient_key.pem without password.
And also can make a apiclient_cert.p12 without password.
And also can send a request to the website with java or curl.
But I still don't know what can I do in openresty enviroments.
The point to implement the function
1、With resty.http can do it?
2、if resty.http can not do it , which api of openresty can do it?

thanks

**Comment by detailyang** [comment]
@walterdev 
I'm so sorry that i misunderstanrd what you says 
>but seemingly can not be loaded with a SSL certificate with password

Actually, the `https://api.mch.weixin.qq.com/secapi/pay/refund` want to mutual authentication between from the server and client. But now the cosocket in openresty does not support this feature:(

**Comment by walterdev** [comment]
@detailyang 

> Actually, the https://api.mch.weixin.qq.com/secapi/pay/refund want to do mutual authentication between from the server and client. But now the cosocket in openresty does not support this feature:(

I'm so sad to hear that. :(
OK, I will invoke the canmmand of curl to implement the function.
Thanks for your help.

**Comment by agentzh** [comment]
@walterdev You know that invoking curl (or any other shell commands) directly from within openresty will block the event loop horribly? If you really need to invoke a shell command, then you should at least use something like [lua-resty-shell](https://github.com/juce/lua-resty-shell).

**Comment by walterdev** [comment]
@agentzh 
Yes , I'm being bothered by this , and I want to know more way to solve this problem better . 
Can you give me more help?

**Comment by detailyang** [comment]
@watlerdev

Because you are using the OpenResty, So we have a another way to solve this:) 

We can use the nginx to reverse proxy the `https://api.mch.weixin.qq.com` with mutual authentication as the following:

```bash
server {
    listen 9999;
    
    proxy_ssl_certificate xx.cer;
    proxy_ssl_certificate_key xx.key;
    
    location / {
      proxy_pass https://api.mch.weixin.qq.com;
      proxy_set_header Host api.mch.weixin.qq.com;
    }
}
```

So now you can request the uri `http://localhost:9999/secapi/pay/refund` by resty.http. 

By the way if OpenResty provide the feature for mutual authentication, it's easy to replace uri to `https://api.mch.weixin.qq.com/secapi/pay/refund`.

**Comment by walterdev** [comment]
@detailyang 

proxy_ssl_certificate and proxy_ssl_certificate_key can load the certificate .
but the information of forward proxy has been in my redis. 
It's dynamic data ( include proxy_pass and  proxy_ssl_certificate ,etc ).
and the certificate was loaded by proxy_ssl_certificate when nginx is starting.
I think this config is static.so I want to find a dynamic way to load certificate.

**Comment by detailyang** [comment]
@walterdev 
It looks now we have to let OpenResty support the feature about mutual authentication :)

**Comment by agentzh** [comment]
Well, patches welcome for adding support for the client-side SSL certificates in the cosocket API :)

**Comment by detailyang** [comment]
@agentzh let me try to implement this feature:)
